python/ray/_private/runtime_env.py

```diff
@@ -0,0 +1,298 @@
+import hashlib
+import logging
+import inspect
+
+from filelock import FileLock
+from pathlib import Path
+from zipfile import ZipFile
+from ray.job_config import JobConfig
+from enum import Enum
+from ray.experimental import internal_kv
+from ray.core.generated.common_pb2 import RuntimeEnv
+from typing import List, Tuple
+from types import ModuleType
+from urllib.parse import urlparse
+import os
+import sys
+
+PKG_DIR = "/tmp/ray/runtime_resources"
+
+logger = logging.getLogger(__name__)
+
+FILE_SIZE_WARNING = 10 * 1024 * 1024  # 10MB
+FILE_SIZE_LIMIT = 50 * 1024 * 1024  # 50MB
+
+
+class Protocol(Enum):
+    """A enum for supported backend storage."""
+
+    # For docstring
+    def __new__(cls, value, doc=None):
+        self = object.__new__(cls)
+        self._value_ = value
+        if doc is not None:
+            self.__doc__ = doc
+        return self
+
+    GCS = "gcs", "For package created and managed by system."
+    PIN_GCS = "pingcs", "For package created and managed by users."
+
+
+def _xor_bytes(left: bytes, right: bytes) -> bytes:
+    if left and right:
+        return bytes(a ^ b for (a, b) in zip(left, right))
+    return left or right
+
+
+def _zip_module(path: Path, relative_path: Path, zip_handler: ZipFile) -> None:
+    """Go through all files and zip them into a zip file"""
+    for from_file_name in path.glob("**/*"):
+        file_size = from_file_name.stat().st_size
+        if file_size >= FILE_SIZE_LIMIT:
+            raise RuntimeError(f"File {from_file_name} is too big, "
+                               "which currently is not allowd ")
+        if file_size >= FILE_SIZE_WARNING:
+            logger.warning(
+                f"File {from_file_name} is too big ({file_size} bytes). "
+                "Consider exclude this file in working directory.")
+        to_file_name = from_file_name.relative_to(relative_path)
+        zip_handler.write(from_file_name, to_file_name)
+
+
+def _hash_modules(path: Path) -> bytes:
+    """
```

Comment:
First line of docstring should be one line and on the same line as triple quote. e.g.,
```
"""This is a short description.

Now I'll elaborate...
"""